Async struct method, <_'> lifetime shadowing

Hello, I ran into a problem using `<'_>` lifetime with async struct methods, here is the example:

    struct Foo<'a> { swag: &'a i32 }

    impl Foo<'_> {
        async fn bar(&self) -> i32 {
            1337
        }
    }

The error is:
`error[E0496]: lifetime name `'_` shadows a lifetime name that is already in scope`

Everything is fine if I use explicit lifetime like `'a`.

Version: 1.39 Nighly (2019-09-19 9b9d2aff8de4d499b4ba on playground)

--- a/src/librustc/hir/lowering.rs
+++ b/src/librustc/hir/lowering.rs
@@ -844,7 +844,7 @@ impl<'a> LoweringContext<'a> {
     /// header, we convert it to an in-band lifetime.
     fn collect_fresh_in_band_lifetime(&mut self, span: Span) -> ParamName {
         assert!(self.is_collecting_in_band_lifetimes);
-        let index = self.lifetimes_to_define.len();
+        let index = self.lifetimes_to_define.len() + self.in_scope_lifetimes.len();
         let hir_name = ParamName::Fresh(index);
         self.lifetimes_to_define.push((span, hir_name));
         hir_name

--- a/src/test/ui/async-await/async-assoc-fn-anon-lifetimes.rs
+++ b/src/test/ui/async-await/async-assoc-fn-anon-lifetimes.rs
@@ -1,0 +1,23 @@
+// check-pass
+// Check that the anonymous lifetimes used here aren't considered to shadow one
+// another. Note that `async fn` is different to `fn` here because the lifetimes
+// are numbered by HIR lowering, rather than lifetime resolution.
+
+// edition:2018
+
+struct A<'a, 'b>(&'a &'b i32);
+struct B<'a>(&'a i32);
+
+impl A<'_, '_> {
+    async fn assoc(x: &u32, y: B<'_>) {
+        async fn nested(x: &u32, y: A<'_, '_>) {}
+    }
+
+    async fn assoc2(x: &u32, y: A<'_, '_>) {
+        impl A<'_, '_> {
+            async fn nested_assoc(x: &u32, y: B<'_>) {}
+        }
+    }
+}
+
+fn main() {}

